### PR TITLE
[#23] [Backend] As a user, I can submit my survey response

### LINKED ISSUE #23: [Backend] As a user, I can submit my survey response
## Why

At the final step, users must be able to submit their responses. The mobile app must show the Submit button instead of the Next button to inform users that the app will submit their response on the next action.

## Acceptance Criteria

Call `POST /api/v1/responses` endpoint with the following body:

```
{
  "survey_id": "{{survey-id}}",
  "questions": [
    {{question-id-with-answers}},
    ...
  ]
}
```

For `question-id-with-answers`, there are two possible types of this data structure:

1. With ID
```json
{
"id": "{{question-id}}",
"answers": [
 {
   "id": "{{answer-id}}"
 }
]
}
```

2. With ID and Answer
```json
{
"id": "{{question-id}}",
"answers": [
 {
   "id": "{{answer-id}}",
   "answer": "{{answer}}"
 }
]
}
```

## Resources

[Postman - Nimble Survey API](https://documenter.getpostman.com/view/11835486/T1LJjTqj#8de38274-3f9f-4b27-8122-58d7399b4c69)

### PR DESCRIPTION
**Pull request by Wadeewee**
#23

## What happened 👀

Call `POST /api/v1/responses` endpoint with the following body:

```
{
  "survey_id": "{{survey-id}}",
  "questions": [
    {{question-id-with-answers}},
    ...
  ]
}
```

For `question-id-with-answers`, there are two possible types of this data structure:

1. With ID
```json
{
"id": "{{question-id}}",
"answers": [
 {
   "id": "{{answer-id}}"
 }
]
}
```

2. With ID and Answer
```json
{
"id": "{{question-id}}",
"answers": [
 {
   "id": "{{answer-id}}",
   "answer": "{{answer}}"
 }
]
}
```

## Insight 📝

- Create `SubmitSurveyAnswerRequest`, `SubmitSurveyQuestionRequest`, and `SubmitSurveyRequest`.
- Create `SubmitSurveyAnswerModel`, and `SubmitSurveyQuestionModel`.
- Create `SubmitSurveyUseCase`.
- Add method `submitSurvey` in `survey_repository`.

## Proof Of Work 📹

```
flutter: *** Request ***
flutter: uri: https://survey-api.nimblehq.co/api/v1/responses
flutter: method: POST
flutter: responseType: ResponseType.json
flutter: followRedirects: true
flutter: persistentConnection: true
flutter: connectTimeout: 0:00:03.000000
flutter: sendTimeout: null
flutter: receiveTimeout: 0:00:05.000000
flutter: receiveDataWhenStatusError: true
flutter: extra: {}
flutter: headers:
flutter:  Content-Type: application/json; charset=utf-8
flutter:  Authorization: Bearer mR97XGClZCnjCkbQkOmgZii2bQwzLOa4Z9abEAvUaBg
flutter:  content-length: 227
flutter: data:
flutter: {survey_id: d5de6a8f8f5f1cfe51bc, questions: [Instance of 'SubmitSurveyQuestionRequest', Instance of 'SubmitSurveyQuestionRequest']}
flutter:
flutter: *** Response ***
flutter: uri: https://survey-api.nimblehq.co/api/v1/responses
flutter: statusCode: 201
flutter: headers:
flutter:  connection: keep-alive
flutter:  cache-control: max-age=0, private, must-revalidate
flutter:  date: Tue, 06 Jun 2023 12:37:09 GMT
flutter:  vary: Accept-Encoding, Origin
flutter:  content-encoding: gzip
flutter:  referrer-policy: strict-origin-when-cross-origin
flutter:  report-to: {"endpoints":[{"url":"https:\/\/a.nel.cloudflare.com\/report\/v3?s=maF9DDjjBirLYVne5dVET0nMshzNMZHX7CeYxU%2F3O7M44de6pR6vyHKWKtvh%2BVJx33Zym6t6GkN2hZof5Kl2PJeF%2FXVm8VbgMaFvvvcZkUAF1JrrNZtU87IN5IHcu1vI24tr5kLTZjZx"}],"group":"cf-nel","max_age":604800}
flutter:  cf-cache-status: DYNAMIC
flutter:  x-permitted-cross-domain-policies: none
flutter:  content-type: application/json; charset=utf-8
flutter:  x-xss-protection: 0
flutter:  server: cloudflare
flutter:  x-request-id: b3795f5b-f251-4f0b-b384-77f3e3998248
flutter:  alt-svc: h3=":443"; ma=86400
flutter:  content-length: 28
flutter:  nel: {"success_fraction":0,"report_to":"cf-nel","max_age":604800}
flutter:  x-download-options: noopen
flutter:  cf-ray: 7d30b47d49e226f5-BKK
flutter:  x-runtime: 0.016270
flutter:  etag: W/"9c9270275739c7283745bb5c0f381983"
flutter:  via: 1.1 vegur
flutter:  x-frame-options: SAMEORIGIN
flutter:  x-content-type-options: nosniff
flutter: Response Text:
flutter: {}
```

### CODE REVIEW

File: lib/model/submit_survey_question_model.dart [hunk 1]
```diff
@@ -0,0 +1,15 @@
+import 'package:equatable/equatable.dart';
+import 'package:survey_flutter_ic/model/submit_survey_answer_model.dart';
+
+class SubmitSurveyQuestionModel extends Equatable {
```
Comment (Wadeewee) [inline lib/model/submit_survey_question_model.dart:4]: TODO Remove `SubmitSurveyQuestionModel` and use `SurveyQuestionModel` instead.

File: lib/model/submit_survey_answer_model.dart [hunk 2]
```diff
@@ -0,0 +1,14 @@
+import 'package:equatable/equatable.dart';
+
+class SubmitSurveyAnswerModel extends Equatable {
```
Comment (Wadeewee) [inline lib/model/submit_survey_answer_model.dart:3]: TODO, Use `SurveyAnswerModel` instead of `SubmitSurveyAnswerModel`.

Comment (Wadeewee) [reply lib/model/submit_survey_answer_model.dart:3]: To avoid confusion I think it better to keep it.

**Review by manh-t — APPROVED** [review]

File: lib/usecase/submit_survey_use_case.dart [hunk 3]
```diff
@@ -0,0 +1,22 @@
+import 'package:injectable/injectable.dart';
+import 'package:survey_flutter_ic/api/exception/network_exceptions.dart';
+import 'package:survey_flutter_ic/api/repository/survey_repository.dart';
+import 'package:survey_flutter_ic/api/request/submit_survey_request.dart';
+import 'package:survey_flutter_ic/usecase/base/base_use_case.dart';
+
+@Injectable()
+class SubmitSurveyUseCase extends UseCase<void, SubmitSurveyRequest> {
```
Comment (hoangnguyen92dn) [inline lib/usecase/submit_survey_use_case.dart]: @Wadeewee The `SubmitSurveyRequest` should be used in Repository or Service layer instead. In the UseCase, we should have a new `Input` model for the UseCase's param

**Review by hoangnguyen92dn — CHANGES_REQUESTED** [review]

File: lib/usecase/submit_survey_use_case.dart (hunk 3 above)
Comment (Wadeewee) [reply lib/usecase/submit_survey_use_case.dart]: @hoangnguyen92dn, May I know why we should not map `model` to `request` since in `VM` 🙏

Comment (hoangnguyen92dn) [reply lib/usecase/submit_survey_use_case.dart]: @Wadeewee Technically, the request/response models are supposed to interact with the backend side which contains `toJson` or `fromJson` for API usage, in the MVVM Clean architecture, the Repository layer is in charge of handling data with data source (Remote or Local), that's why avoid exposing the Request models in VM and UseCase. Similar to this: https://github.com/Wadeewee/survey-flutter-ic-pooh/blob/372cca960afb6083a7b5ff035bd0d8886184eb6c/lib/api/repository/auth_repository.dart#L36-L44

Comment (Wadeewee) [reply lib/usecase/submit_survey_use_case.dart]: > Technically, the request/response models are supposed to interact with the backend side which contains `toJson` or `fromJson` for API usage, in the MVVM Clean architecture, the Repository layer is in charge of handling data with data source (Remote or Local),

I agreed with this 🤝  @hoangnguyen92dn

**Review by hoangnguyen92dn — APPROVED** [review]

**Pull request merged**